### PR TITLE
Fixes #114 : Add `neuron_compiler` and `neuron_compiler_version` to c…

### LINKED ISSUE #114: Add neuronx-cc or neuron-cc version to config.json
We noticed that compiled models might not be compatibel across neuron-cc or neuronx-cc versions and raise runtime errors, e.g. 
```bash
File /usr/local/lib/python3.8/dist-packages/torch/jit/_serialization.py:162, in load(f, map_location, _extra_files)
    160 cu = torch._C.CompilationUnit()
    161 if isinstance(f, str) or isinstance(f, pathlib.Path):
--> 162     cpp_module = torch._C.import_ir_module(cu, str(f), map_location, _extra_files)
    163 else:
    164     cpp_module = torch._C.import_ir_module_from_buffer(
    165         cu, f.read(), map_location, _extra_files
    166     )

RuntimeError: Could not load the model status=2 message=Invalid
```
Those provide not the best UX. We should add the compiled neuron version into the config.json as `neuron_version` and assert those when using the NeuronModel class.

### PR DESCRIPTION
**Pull request by dangdana**
…onfig.json

### CODE REVIEW

**Comment by philschmid** [comment]
Thank you for opening the PR! We ll look at it. 
This should close then #114

File: optimum/neuron/modeling_base.py [hunk 1]
```diff
@@ -154,6 +156,16 @@ def _from_pretrained(
             else:
                 file_name = neuron_files[0].name
 
+        # check compiler compatability of saved model vs system installation
+        # Vagrantly assumes that the user won't try to load an xcc on a cc machine (and vise/versa)
```
Comment (JingyaHuang) [inline optimum/neuron/modeling_base.py:160]:
```suggestion
        # Check compiler compatibility(compiler type and version) of the saved model vs. system.
```

File: optimum/neuron/modeling_base.py [hunk 2]
```diff
@@ -154,6 +156,16 @@ def _from_pretrained(
...
+        if hasattr(config, "neuron_compiler") and hasattr(config, "neuron_compiler_version"):
```
Comment (JingyaHuang) [inline optimum/neuron/modeling_base.py:161]: Here I would suggest checking the compiler type(neuronx-cc / neuron-cc) and the version separately.
Like
* First check if the compiler type matches, if not raise.
* Then check if the version is good.

File: optimum/neuron/modeling_base.py [hunk 3]
```diff
@@ -154,6 +156,16 @@ def _from_pretrained(
...
+            min_version = get_neuronxcc_version() if config.neuron_compiler == "neuronx-cc" else get_neuroncc_version()
+            if version.parse(config.neuron_compiler_version) > version.parse(min_version):
```
Comment (JingyaHuang) [inline optimum/neuron/modeling_base.py:163]: Here I am curious about the backward compatibility of the neuron compiler. Would a model compiled with an older version of neuron(x)-cc compiler be compatible to run under a newer environment?  @mmcclean-aws

File: tests/test_modeling.py [hunk 4]
```diff
@@ -174,6 +174,13 @@ def test_save_model(self):
             self.assertTrue(NEURON_FILE_NAME in folder_contents)
             self.assertTrue(CONFIG_NAME in folder_contents)
 
+    def test_save_load_model(self):
+        # a config.json can change when converted and saved, so check loadability is maintained
+        with tempfile.TemporaryDirectory() as tmpdirname:
```
Comment (JingyaHuang) [inline tests/test_modeling.py:179]: Here I am not sure fully understood the test.

Are you worried that there might be a problem with re-loading the config if we save it once again? (for exemple in the case of this PR, users previously saved their models can run `model.save_pretrained()` to update their config with compiler version info)

If so, I think it is good to think of it but not really mandatory to add this test, as `NeuronBaseModel` is designed to be quite tolerant when having extra or lacking of not mandatory information in the config file.

**Review by JingyaHuang — COMMENTED** [review]
Hey @dangdana, thanks a lot for contributing, the PR looks awesome 💯  !

Left a few nits to discuss.

**Comment by JingyaHuang** [comment]
And to fix the styling, you can run 
```bash
make style
```

File: tests/test_modeling.py (hunk 4 above)
Comment (dangdana) [reply tests/test_modeling.py:179]: I used a test like this to ensure that the new config.json fields were being correctly processed when the NeuronBaseModel was loaded.  It did catch a bug where I had ">=" on the version check, when I should have just had ">".  

You're right about the tolerances of NeuronBaseModel.  I think this test served its purpose and is unlikely to catch anything in the future.  I'll remove it.

File: optimum/neuron/modeling_base.py (hunk 3 above)
Comment (dangdana) [reply optimum/neuron/modeling_base.py:162]: NEFFs are backwards compatible, but official confirmation would be good from @mcclean-aws .

Comment (mmcclean-aws) [reply optimum/neuron/modeling_base.py:176]: Yes, the Neuron runtime ensures backwards compatibility of NEFFs

**Review by JingyaHuang — APPROVED** [review]
LGTM, thanks again for the PR! Let's get it merged! @dangdana

**Pull request merged**